libr/cons/canvas.c

```diff
@@ -310,42 +379,54 @@ R_API void r_cons_canvas_write(RConsCanvas *c, const char *s) {
 }
 
 R_API char *r_cons_canvas_to_string(RConsCanvas *c) {
-	int x, y, olen = 0;
+	int x, y, olen = 0, attr_x = 0;
 	char *o;
-	const char *b;
 	const char **atr;
-	int is_first = true;
+	int is_first = true, last_nonwhite = 0;
 
 	if (!c) {
 		return NULL;
 	}
-	b = c->b;
-	o = calloc (1, (c->w * (c->h + 1)) * (CONS_MAX_ATTR_SZ));
+
+	for (y = 0; y < c->h; y++) {
+		olen += c->blen[y] + 1;
+	}
+	o = calloc (1, olen * (CONS_MAX_ATTR_SZ));
 	if (!o) {
 		return NULL;
 	}
+
+	olen = 0;
 	for (y = 0; y < c->h; y++) {
 		if (!is_first) {
 			o[olen++] = '\n';
 		}
 		is_first = false;
-		for (x = 0; x < c->w; x++) {
-			const int p = x + (y * c->w);
-			atr = attr_at (c, p);
-			if (atr && *atr) {
-				strcat (o, *atr);
-				olen += strlen (*atr);
+		attr_x = 0;
+		for (x = 0; x < c->blen[y]; x++) {
+			if ((c->b[y][x] & 0xc0) != 0x80) {
+				atr = attr_at (c, y*c->w + attr_x);
+				if (atr && *atr) {
+					strcat (o, *atr);
```

Comment:
strcat already does strlen, so the strlen after that is unnecessary, so you can do strlen + memcpy(or memmove if its in the same buffer) instead